IPv6: support link-local addresses

bajtos opened the pull request:
Modified uv_ip6_addr() to fill sin6_scope_id for link-local addresses (Linux and Windows only).

Fixes #271

bajtos committed:
linux,darwin,win: link-local IPv6 addresses

Modified uv_ip6_addr() to fill sin6_scope_id for link-local addresses.

Fixes #271

bnoordhuis commented on src/uv-common.c:
```diff
@@ -170,16 +173,41 @@ struct sockaddr_in uv_ip4_addr(const char* ip, int port) {
   return addr;
 }
 
-
```
Style: two newlines between functions (i.e. don't remove this line.)

bnoordhuis commented on src/uv-common.c:
```diff
@@ -170,16 +173,41 @@ struct sockaddr_in uv_ip4_addr(const char* ip, int port) {
 struct sockaddr_in6 uv_ip6_addr(const char* ip, int port) {
   struct sockaddr_in6 addr;
+  char * address_part = NULL;
+  const char* zone_index = NULL;
```
Style: split declaration and assignment.

bnoordhuis commented on src/uv-common.c:
```diff
@@ -170,16 +173,41 @@ struct sockaddr_in uv_ip4_addr(const char* ip, int port) {
+  }
+#else
+  /* TODO - implement sin6_scope_id for other platforms
+   * Windows IPv6 stack uses numeric zone indexes, e.g., fe80::3%1. The index is determined by the interface number;
```
Style: long line, wrap at 80 columns.

bnoordhuis commented on src/uv-common.c:
```diff
@@ -170,16 +173,41 @@ struct sockaddr_in uv_ip4_addr(const char* ip, int port) {
+#ifdef __linux__
+  zone_index = strchr(ip, '%');
+  if (zone_index != NULL) {
+    address_part = strndup(ip, zone_index - ip);
```
Don't use a dynamic allocation here, we try to avoid that wherever possible in libuv. Use an on-stack buffer that is large enough to hold the address.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,89 @@
+        address[i*5 + j] = address[i*4 + j];
+      if (i > 0)
+        address[i*5 - 1] = ':';
+    }
```
Style: indentation.

bajtos replied:
I don't understand which part is wronly indented? The function will be removed, but I'd like to know the correct style.

bnoordhuis replied:
All of it. :-) Use two spaces, not four.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,89 @@
+
+void stringify_address(char* address) {
+    int i, j;
+    ASSERT(strlen(address)==32);
```
Style: spaces around the ==.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,89 @@
+
+typedef void (*iface_info_cb)(const char* ip6_addr, const char* device_name, int iface_index, int scope_value);
+
+void stringify_address(char* address) {
```
Style: two newlines between functions and the top declarations.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,89 @@
+    }
+}
+
+void foreach_network_interface(iface_info_cb iface_cb) {
```
There is a libuv function for that: uv_interface_addresses()

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,89 @@
+  foreach_network_interface(&test_ip6_addr_scope);
+  return 0;
+}
+
```
Style: no newline at EOF.

bnoordhuis commented on build.mk:
```diff
@@ -136,6 +136,7 @@ TESTS= \
 	test/test-udp-send-and-recv.o \
 	test/test-util.o \
 	test/test-walk-handles.o \
+	test/test-ip6-addr.o \
```
They're sorted alphabetically (but not in uv.gyp for some reason. You can leave that one as-is.)

bnoordhuis commented on src/uv-common.c:
```diff
@@ -170,14 +173,38 @@ struct sockaddr_in uv_ip4_addr(const char* ip, int port) {
+#if defined(__linux__) || defined(_WIN32)
+  zone_index = strchr(ip, '%');
+  if (zone_index != NULL) {
+    assert(zone_index - ip < 40);
```
sizeof(address_part)

bnoordhuis commented on src/uv-common.c:
```diff
@@ -170,14 +173,38 @@ struct sockaddr_in uv_ip4_addr(const char* ip, int port) {
+  zone_index = strchr(ip, '%');
+  if (zone_index != NULL) {
+    assert(zone_index - ip < 40);
+    strncpy(address_part, ip, zone_index - ip);
```
Doesn't zero-terminate the string when the address is > sizeof(address_part). Don't rely on the assert() to catch that.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,83 @@
+#endif
+
+typedef void (*iface_info_cb)(const char* ip6_addr, const char* device_name,
+     unsigned iface_index);
```
Style: when arguments are split over multiple lines, they should line up (and one argument per line.)

bajtos replied:
This rule is not consistently obeyed in the code base (see e.g. uv_read2_cb in include/uv.h), but I will fix it per your suggestion.

piscisaureus replied:
Agree with @bajtos. I use the double-indent style for function signatures (but not function calls) too.

bajtos replied:
@piscisaureus: thanks :)

I'll leave it as it is for now, we can always change it later.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,83 @@
+typedef void (*iface_info_cb)(const char* ip6_addr, const char* device_name,
+     unsigned iface_index);
+
+void call_iface_info_cb(iface_info_cb iface_cb, char const* iface_name, struct sockaddr_in6 const* address) {
```
Long line, wrap at 80 columns. Happens in a few other places too.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,83 @@
+  result = uv_interface_addresses(&addresses, &count);
+  ASSERT(result.code == UV_OK);
+
+  for (ix=0; ix<count; ix++) {
```
Style: spaces around assignment/comparison operators.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,83 @@
+
+  for (ix=0; ix<count; ix++) {
+    if (addresses[ix].address.address4.sin_family != AF_INET6) continue;
+    call_iface_info_cb(iface_cb, addresses[ix].name, &addresses[ix].address.address6);
```
Merge that function? It's only used here.

EDIT: I guess you could just inline everything into TEST_IMPL(ip6_addr_link_local).

bajtos replied:
I disagree on this point. I put the code in a standalone function so that the test method is more readable. The same applies for call_iface_info_cb(). I strongly believe that many small functions make better/cleaner code, which is also easier to maintain.

Why would you want to inline/merge the function?

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,83 @@
+  iface_cb(string_address, iface_name, address->sin6_scope_id);
+}
+
+void foreach_ip6_interface(iface_info_cb iface_cb) {
```
Style: two newlines between functions.

bnoordhuis commented on test/test-list.h:
```diff
@@ -204,6 +204,9 @@
 TEST_DECLARE   (poll_duplex)
 TEST_DECLARE   (poll_unidirectional)
 TEST_DECLARE   (poll_close)
+#if defined(__linux__) || defined(_WIN32)
```
Preferably don't #ifdef things in test-list.h. Do it in the test itself and return 0 when we're on an unsupported platform (with maybe a printed message telling the user the test is skipped.)

bajtos replied:
OK, I'll do the ifdef in the test method. The reason for ifdefing in test-list was that there are few other tests turned on/off this way.

BTW, I don't think there is an easy way how to print "test skipped" message. Output of LOG macro is printed only when the test fails.

bnoordhuis commented on test/test-ip6-addr.c:
```diff
@@ -0,0 +1,83 @@
+  int count, ix;
+  uv_interface_address_t* addresses;
+
+  result = uv_interface_addresses(&addresses, &count);
```
Call uv_free_interface_addresses() when you're done with it.

bajtos replied:
This has been fixed in the last commit (github did not pick the change in this diff view).

bnoordhuis commented on src/uv-common.c:
```diff
@@ -170,14 +173,38 @@ struct sockaddr_in uv_ip4_addr(const char* ip, int port) {
+#elif defined(_WIN32)
+  return atoi(zone_index);
+#endif
+}
```
I suggest you inline this into the `#if defined(__linux__) || defined(_WIN32)` block further down below. It generates 'unused function' and 'no return value' warnings on non-Linux/Windows platforms.

bajtos replied:
I don't like that idea because the zone index handling is already awfully long. I defined a new macro `UV_PLATFORM_HAS_IP6_LINK_LOCAL_ADDRESS` and protected the helper function with `#ifdef` instead.

Perhaps we could move `uv__parse_zone_index` to platform-specific files (e.g. [include/uv-private/uv-win.h](https://github.com/joyent/libuv/blob/master/include/uv-private/uv-win.h)) in order to keep [src/uv-common.c](https://github.com/joyent/libuv/blob/master/src/uv-common.c) cleaner, possibly defining the helper function as a macro instead?

What do you think?

bnoordhuis commented:
/cc @piscisaureus

bajtos commented:
Hello, what's the status of this pull request? What changes are necessary to finish this work and have the request pulled to master?

saghul commented:
Looks like if_nametoindex is also supported on OSX. https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man3/if_nametoindex.3.html not sure about other BSDs though.

bajtos commented:
I fixed remaining issues, added Darwin implementation and squashed the changes into a single commit.

@bnoordhuis @piscisaureus could you please review?

bnoordhuis commented:
Landed on July 4 in fd45f87.
